### ISSUE TITLE
Option to add CORS config or expose outside traffic when running in Docker

### ISSUE THREAD

**Issue by silvaDominic** [jamespfennell/transiter#143]
## Issue
Can't access transiter resources from other port.

## Expected
Can query API from client side app.

## Actual
Can't query API from client side app -- get typical CORS error:
```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:8080/systems/us-ny-subway. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.
```

## Context:
I wanted to throw together a simple web applet (using Vite) to try out the API and immediately ran into this problem when running with docker. I could obviously download the project and setup my own CORs policy or choose to expose to outside traffic, but was wondering if not having the option to control this was intended. This is outside my domain so forgive me if this behavior implied/typical.

## Additional Notes:
Consider adding a "Discussion" section to the repo for concerns like this.

**Comment by jamespfennell** [comment]
Hey! I'm not sure that setting CORS headers for localhost development will actually work. My understanding (correct me if I'm wrong) is that browsers don't honor CORS headers for localhost so you'll still have the same issue. This is for security reasons. If I'm wrong, we can work something out.

To unblock you, I just created a new domain `cors.transiter.dev` that is the same as the demo site but has `access-control-allow-origin: *`. It has the NYC subway installed, under ID `us-ny-subway`. Is that useful? I'm intending this for prototyping; if you want to deploy an app online you should run your own Transiter.

**Comment by silvaDominic** [comment]
Hey @jamespfennell thanks for getting back!

> My understanding (correct me if I'm wrong) is that browsers don't honor CORS headers for localhost so you'll still have the same issue.

Not sure where you came to understand that. There could be issues if you mix usage of localhost and 127.0.0.1 across client and server, but otherwise the client will respect the headers.

I appreciate you spinning up another instance to get me unblocked and can confirm that works but I was able to perform a workaround using [Vite's proxy feature](https://vite.dev/config/server-options#server-proxy). Otherwise, I will have to deal with this when I deploy a production environment.

Also, do you mind sharing what lib you used for cors setup and maybe some sample code of how you configured it in the project? I have 0 exp with Go and tried in vain to get something working yesterday.

**Comment by jamespfennell** [comment]
Thanks for the localhost correction! I must look more into it and then add some docs about doing local development because I've encountered this CORS issue exactly myself.

> Also, do you mind sharing what lib you used for cors setup and maybe some sample code of how you configured it in the project? I have 0 exp with Go and tried in vain to get something working yesterday.

In general Transiter should be deployed behind a reverse proxy like Caddy or nginx. (This is [described in the docs](https://docs.transiter.dev/deployment/#reverse-proxy) but it's sorta brief). HTTP requests go to the reverse proxy first, which hands them off to Transiter, which responds to the proxy and then the proxy responds to the user. You can configure the reverse proxy so that it sets any HTTP headers you want. This has nothing to do with Go and is just using the fact that Transiter exports a vanilla HTTP API.

For example, I use Caddy and added the following Caddyfile configuration for the new domain I set up. On my cloud VM Transiter is running on `127.0.0.1:8010`, but can't be contacted directly from the internet. Note the `header_down` directive which adds the CORS header:

```
cors.transiter.dev {
    encode gzip
    reverse_proxy 127.0.0.1:8010 {
        header_up X-Transiter-Host "https://cors.transiter.dev"
        header_down Access-Control-Allow-Origin "*"
    }
}
```

**Comment by silvaDominic** [comment]
Yeah I did see that note, but was trying to avoid more setup. Seems like this might be simple still and I gotta do it anyway so I'll give it a shot.

Otherwise, I think we can resolve this issue. Only thing I can think of is maybe including some more notes about local development like you said.

And great work on this tool btw! It's the nicest transit tool I've come across. I'll reach out to you once I finished what I'm working on.

**Comment by jamespfennell** [comment]
I thought/investigated a little more, and have 2 comments.

First, I put together a Docker compose file that puts Transiter behind simple reverse proxy (using Caddy) and sets the CORS header. Here it is:

```
version: '3.5'

services:

  transiter:
    image: jamespfennell/transiter:latest
    ports:
      - "127.0.0.1:8080:8080"
      - "127.0.0.1:8081:8081"
      - "127.0.0.1:8082:8082"
      - "127.0.0.1:8083:8083"
    restart: always
    command:
      - server
      - --log-level
      - info
      - --postgres-connection-string
      - "postgres://transiter:transiter@postgres:5432/transiter"

  caddy:
    image: caddy:2.8.4
    ports:
      - "127.0.0.1:8090:8090"
    command:
      - caddy
      - reverse-proxy
      - --debug
      - --from
      - :8090
      - --to
      - transiter:8080
      - --header-down
      - "Access-Control-Allow-Origin: *"

  postgres:
    image: postgis/postgis:14-3.4
    environment:
      - POSTGRES_USER=transiter
      - POSTGRES_PASSWORD=transiter
      - POSTGRES_DB=transiter
    restart: always
```

Here the Transiter public HTTP API listens on port 8080, but you can make requests through the reverse proxy on port 8090 instead, and if you do the CORS header will be set because of the `--header-down` flag on the proxy.

I probably should put this compose file in the docs.

Second, I realized I don't actually hit the CORS issue when I'm developing my own app because it's built using React and I use [this React proxy feature](https://create-react-app.dev/docs/proxying-api-requests-in-development/) to proxy API requests to my prod URL. It proxies them in a way that doesn't require CORS to be set because the prod URL API requests themselves aren't made by the browser.

**Comment by cedarbaum** [comment]
Coincidentally I also was dealing with configuring CORS recently! Based on this conversation and @jamespfennell's sample Caddy configs above, I went ahead and sent a PR to update the documentation (@jamespfennell if you were already working on this and have a draft, feel free to use that instead and I can close the PR): https://github.com/jamespfennell/transiter/pull/145

I also added some documentation links for configuring proxies for local web development (e.g., using Vite, Create React App, and Next.js) per @silvaDominic's suggestion. If the PR looks good to you both, I can merge it and close this issue.

**Comment by cedarbaum** [comment]
I am going to close this for now since we've updated the docs to discuss handling common CORS issues. @silvaDominic, if you're stilling having any issues, feel free to reopen.
